samples/RoutingSample.Web/DomainPolicy/DomainMatcherPolicy.cs

```diff
@@ -0,0 +1,184 @@
+
+                            if (edgeKey == domain)
+                            {
+                                kvp.Value.Add(endpoint);
+                                break;
+                            }
+                        }
+                    }
+                }
+            }
+
+            return edges
+                .Select(kvp => new PolicyNodeEdge(kvp.Key, kvp.Value))
+                .ToArray();
+        }
+
+        public PolicyJumpTable BuildJumpTable(int exitDestination, IReadOnlyList<PolicyJumpTableEdge> edges)
+        {
+            if (edges == null)
+            {
+                throw new ArgumentNullException(nameof(edges));
+            }
+
+            // Since our 'edges' can have wildcards, we do a sort based on how wildcard-ey they
+            // are then then execute them in linear order.
+            var ordered = edges
+                .Select(e => (domain: (HostString)e.State, destination: e.Destination))
+                .OrderBy(e => GetScore(e.domain))
+                .ToArray();
+
+            return new DomainPolicyJumpTable(exitDestination, ordered);
+        }
+
+        private int GetScore(in HostString domain)
+        {
+            // Higher score == lower priority.
+            if (!string.IsNullOrEmpty(domain.Host) && domain.Port != null)
+            {
+                return 1; // Has host AND port
+            }
+            else if (!string.IsNullOrEmpty(domain.Host) || domain.Port != null)
+            {
+                return 2; // Has host OR port
+            }
+            else
+            {
+                return 3; // Has neither
+            }
+        }
+
+        private class DomainMetadataEndpointComparer : EndpointMetadataComparer<DomainMetadata>
+        {
+            protected override int CompareMetadata(DomainMetadata x, DomainMetadata y)
+            {
+                // Ignore the metadata if it has an empty list of IP endpoint types.
+                return base.CompareMetadata(
+                    x?.Domains.Count > 0 ? x : null,
+                    y?.Domains.Count > 0 ? y : null);
+            }
+        }
+
+        private class DomainPolicyJumpTable : PolicyJumpTable
+        {
+            private (HostString domain, int destination)[] _destinations;
+            private int _exitDestination;
+
+            public DomainPolicyJumpTable(int exitDestination, (HostString domain, int destination)[] destinations)
+            {
+                _exitDestination = exitDestination;
+                _destinations = destinations;
+            }
+
+            public override int GetDestination(HttpContext httpContext)
+            {
+                var destinations = _destinations;
+                for (var i = 0; i < destinations.Length; i++)
+                {
+                    if ((string.IsNullOrEmpty(destinations[i].domain.Host) ||
+                        httpContext.Request.Host.Host == destinations[i].domain.Host) &&
```

Comment:
maybe I misunderstand what you're saying about wildcards? I kinda expected to see handling for something like `*.example.com`

Reply:
By wildcards I meant that it supports *:port, or host:*. I didn't consider wildcards inside the host. That would be a required feature. Will add!